ikarus/controlroutines/pathfollowingtechnique.hh

```diff
@@ -16,54 +18,86 @@
 
 namespace Ikarus {
 
-  struct PathFollowingInformation {
+  struct ControlInformation {
     bool success{false};
+    std::vector<Ikarus::NonLinearSolverInformation> solverInfos{};
+    int totalIterations;
   };
 
-  template <typename NonLinearSolver, typename PathFollowingType = Ikarus::StandardArcLength>
-  requires Concepts::PathFollowingStrategy<PathFollowingType, typename NonLinearSolver::NonLinearOperator>
-  class PathFollowing : public IObservable<ControlMessages> {
+  template <typename NonLinearSolver, typename PathFollowingType = Ikarus::StandardArcLength,
+            typename AdaptiveStepSizing = Ikarus::DefaultAdaptiveStepSizing<>>
+  requires(Concepts::PathFollowingStrategy<PathFollowingType, typename NonLinearSolver::NonLinearOperator,
+                                           Ikarus::SubsidiaryArgs>)
+      || (Concepts::AdaptiveStepSizingStrategy<
+          AdaptiveStepSizing, Ikarus::NonLinearSolverInformation, Ikarus::SubsidiaryArgs,
+          std::remove_cvref_t<typename NonLinearSolver::NonLinearOperator>>)class PathFollowing
+      : public IObservable<Ikarus::ControlMessages> {
   public:
     PathFollowing(const std::shared_ptr<NonLinearSolver>& p_nonLinearSolver, int loadSteps, double stepSize,
-                  PathFollowingType p_pathFollowingType = Ikarus::StandardArcLength{})
+                  PathFollowingType p_pathFollowingType   = Ikarus::StandardArcLength{},
+                  AdaptiveStepSizing p_adaptiveStepSizing = {})
         : nonLinearSolver{p_nonLinearSolver},
           loadSteps_{loadSteps},
           stepSize_{stepSize},
-          pathFollowingType_{p_pathFollowingType} {}
+          pathFollowingType_{p_pathFollowingType},
+          adaptiveStepSizing{p_adaptiveStepSizing} {}
 
-    PathFollowingInformation run() {
-      spdlog::info("Started path following with subsidiary equation: {}", pathFollowingType_.name);
-      PathFollowingInformation info({false});
+    ControlInformation run() {
+      ControlInformation info({false});
       auto& nonOp = nonLinearSolver->nonLinearOperator();
-      this->notify(ControlMessages::CONTROL_STARTED);
+      this->notify(ControlMessages::CONTROL_STARTED, pathFollowingType_.name);
+      using ResidualType = NonLinearSolver::ResidualType;
+      using JacobianType = NonLinearSolver::JacobianType;
+
+      static_assert(
+          not(std::is_same_v<
+                  ResidualType,
+                  double> and ((Ikarus::Std::isSpecializationTypeAndNonTypes<Eigen::Matrix, JacobianType>::value) or (Ikarus::Std::isSpecializationTypeNonTypeAndType<Eigen::SparseMatrix, JacobianType>::value))),
+          "The value of the nonlinear operator cannot be a double while its derivative is a vector or a matrix.");
```

Comment:
The non-linear solver allowed for this controlroutine here makes strong assumptions on the nonlinear solver ,i.e. it always has to be NewtonRaphsonWith Sub args. Thus, maybe hardcode it here, or write a template which checks that the nonlinearsolver we get here hast all the correct things.

Reply:
You can also use the boolean `isLinearSolver` of the Newton method. This should also be renamed accordingly. The boolean basically indicates if this is only a scalar or if there is a LinearSolver, so maybe `hasLinearSolver`would be better which indicates a vector valued problem, i.e. with eigentypes. Can you also write failing checks?